src/Compilers/Test/Utilities/CSharp/CSharpTestBase.cs

```diff
@@ -2112,6 +2113,28 @@ public static IEnumerable<object[]> NonNullTypesTrueAndFalseReleaseDll
         }
         #endregion
 
+        internal static TypeDefinition GetTypeDefinitionByName(MetadataReader reader, string name)
+        {
+            var typeDef = reader.TypeDefinitions.SingleOrDefault(h => reader.StringComparer.Equals(reader.GetTypeDefinition(h).Name, name));
+            if (typeDef.IsNil)
+            {
+                var builder = new StringBuilder();
+                builder.AppendLine();
+                foreach (var typeHandle in reader.TypeDefinitions)
+                {
+                    builder.AppendLine(reader.GetString(reader.GetTypeDefinition(typeHandle).Name));
+                }
+                Assert.True(false, $"Could not find type '{name}'. Found types are: {builder.ToString()}");
+            }
+            return reader.GetTypeDefinition(typeDef);
+        }
+
+        protected static void AssertAttributes(MetadataReader reader, CustomAttributeHandleCollection handles, params string[] expectedNames)
```

Comment:
>AssertAttributes [](start = 30, length = 16)

Why do we need these helpers? Are we reinventing a wheel? Symbols expose information about attributes.

Reply:
I want to verify the attributes in metadata itself, not from source symbols or PE symbols.

---
In reply to: [295437570](https://github.com/dotnet/roslyn/pull/36444#discussion_r295437570) [](ancestors = 295437570)

Reply:
> I want to verify the attributes in metadata itself, not from source symbols or PE symbols.

What is the motivation? Do you not trust PE Symbols to accurately reflect information about attributes? Why use something that nobody knows what it does and how robust it is vs. using well known, well tested API that everybody knows and understands.

---
In reply to: [295492930](https://github.com/dotnet/roslyn/pull/36444#discussion_r295492930) [](ancestors = 295492930,295437570)

Reply:
Although for this attribute most of the trickery is in source symbols, I thought it's good to verify directly. Do you want me to add PE symbol validation as well?
By the way, those two methods aren't new, they are moved from attributes tests.

---
In reply to: [295494332](https://github.com/dotnet/roslyn/pull/36444#discussion_r295494332) [](ancestors = 295494332,295492930,295437570)